File: benchmark/nqueens-vec.py
# N-queens with a fixed-size list (mirroring alcove's vector).
# Same algorithm as nqueens-vec.alc.
def safe(col, qs, row, offset=0):
    while offset < row:
        placed = qs[offset]
        if col == placed: return False
        d = row - offset
        if col - d == placed: return False
        if col + d == placed: return False
        offset += 1
    return True

def try_cols(n, row, col, qs, count):
    while col < n:
        if safe(col, qs, row):
            qs[row] = col
            count += solve(n, row + 1, qs)
        col += 1
    return count

def solve(n, row, qs):
    if row >= n:
        return 1
    return try_cols(n, row, 0, qs, 0)

qs = [0] * 10
print(solve(10, 0, qs))

File: benchmark/nqueens-vec.c
#include <stdio.h>
static int safe(int col, int *qs, int row) {
  for (int o = 0; o < row; o++) { int placed = qs[o]; int d = row - o;
    if (col == placed || col - d == placed || col + d == placed) return 0; }
  return 1;
}
static long long solve(int n, int row, int *qs) {
  if (row >= n) return 1;
  long long count = 0;
  for (int col = 0; col < n; col++)
    if (safe(col, qs, row)) { qs[row] = col; count += solve(n, row + 1, qs); }
  return count;
}
int main(void) { volatile int n = 10; int qs[64] = {0}; printf("%lld\n", solve(n, 0, qs)); return 0; }
